Create 2022-11-29-mroonga.md (#2)
yoshimotoyuk/yoshimotoyuk: 2022-11-29-mroonga.md

* Added support for reference count mode.
  Though this feature can keep fixed memory usage, its performance would be dropped. Thus, memory increments should be first consideration before use.
    <!-- この機能によりメモリー使用量を一定量に保つことができますが、パフォーマンスが悪化します。 そのため、メモリー不足の場合には、この機能を使用する前に、メモリーを増強することを検討してください。-->

  The reference count mode is used with `table_open_cache` <https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_table_open_cache> of MySQL.
    <!-- 参照カウントモードは MySQLの table_open_cache <https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_table_open_cache> とともに使用します。（My SQLのというのがMy SQLのシステム上でという事ならonかinだと思います )-->
  MySQL can cache specified number of tables  with `table_open_cache`. Groonga's object would not be released because the cached tables are still in use.
  <!-- MySQLは table_open_cache で指定した個数のテーブルをキャッシュしておくことができます。キャッシュされているテーブルはまだ使用中なので、Groongaのオブジェクトも解放されません。-->
  
  The tables with low usage would be closed if number of open tables is larger than number specified with `table_open_cache`.
  <!--table_open_cache で指定した個数よりも多いテーブルが開かれたとき、使用頻度が低いテーブルが閉じられます。参照カウントモードが有効なとき、そのタイミングでGroongaのオブジェクトも閉じられます。

> 【吉本】ここの日本語がよくわからなかったです。　Groongaのオブジェクトが開放されるのとオブジェクトが閉じるのは同義ですか？この説明は「どうやってreferencecount modeがメモリー使用量を一定に保つのか？」を説明しているという事であっていますか？-->

Reviewer: > Groongaのオブジェクトが開放されるのとオブジェクトが閉じるのは同義ですか？

ここでは同義です。なので、解放される(release)に統一するのが良いのではないかと思います。

> この説明は「どうやってreferencecount modeがメモリー使用量を一定に保つのか？」を説明しているという事であっていますか？

はい、あっています。

-->
 We need to adjust balance between memory usage and perfomance with value of `table_open_cache` while checking memory usage and value of static variables Open Tables<https://dev.mysql.com/doc/refman/8.0/en/server-status-variables.html#statvar_Open_tables> .     
 <!--ステータス変数の Open_tables <https://dev.mysql.com/doc/refman/8.0/en/server-status-variables.html#statvar_Open_tables> の値とメモリー使用量を見ながら table_open_cache の値を調整することで、メモリー使用量とパフォーマンスのバランスを調整する必要があります。-->

There is how to check `Open_tables` as follows
SHOW GLOBAL STATUS LIKE 'Open_tables';
-- +---------------+-------+
-- | Variable_name | Value |
-- +---------------+-------+
-- | Open_tables   | 643   |
-- +---------------+-------+
Normally, all of tables in use are cached by setting `table_open_cache` larger than `Open Tables`. However, if there is limited memory, memory usage would be decreased by releasing tables with low usage from the cache.
In order to releasing tables with low usage from the cache, `table_open_cache` should be set smaller than `Open Tables`. 
Amount of cache released would scale linearly with how smaller `table_opne_cashe`is. 
As cache is released, memory usage decrease, yet performance become worse.

> <!--通常は Open_tables より table_open_cache を大きくして常に使っているすべてテーブルをキャッシュします。 しかし、メモリーに余裕がない環境では一部の使用頻度の低いテーブルをキャッシュから落として解放することでメモリー使用量を削減します。

Reviewer: キャッシュは、一定数のデータを貯蔵庫に保存しておいて、一定数以上のものがきたら貯蔵庫から外れる（落ちる）というような動きをするのですが、ここでは特にキャッシュから落ちたものはメモリー上から解放される動きをします。

キャッシュから落ちる -> 結果的にメモリー上から解放される

という流れです。

Author: なるほど

  MySQL can cache specified number of tables  with `table_open_cache`. Groonga's object would not be released because the cached tables are still in use.
  <!-- MySQLは table_open_cache で指定した個数のテーブルをキャッシュしておくことができます。キャッシュされているテーブルはまだ使用中なので、Groongaのオブジェクトも解放されません。-->
  
  The tables with low usage would be closed if number of open tables is larger than number specified with `table_open_cache`. Groonga objects would be released at the same time with table closing if the reference count mode is enabled.
  
<!--table_open_cache で指定した個数よりも多いテーブルが開かれたとき、使用頻度が低いテーブルが閉じられます。参照カウントモードが有効なとき、そのタイミングでGroongaのオブジェクトも閉じられます。
【吉本】ここの日本語がよくわからなかったです。　Groongaのオブジェクトが開放されるのとオブジェクトが閉じるのは同義ですか？この説明は「どうやってreferencecount modeがメモリー使用量を一定に保つのか？」を説明しているという事であっていますか？-->

 These are how the reference count mode keep fixed memory usage.

For actual use, we need to adjust balance between memory usage and perfomance with value of `table_open_cache` while checking memory usage and value of static variables Open Tables<https://dev.mysql.com/doc/refman/8.0/en/server-status-variables.html#statvar_Open_tables>. Because this feature can keep fixed memory usage, but its performance would be dropped.

> 

Author: https://github.com/yoshimotoyuk/yoshimotoyuk/pull/2/files#r1034336521
で教えていただいたことと英語の自然な流れで次へ（バランス調整する）をつなげるのに
下記文章たしてみました

For actual use, we need to adjust balance between memory usage and perfomance with value of `table_open_cache` while checking memory usage and value of static variables Open Tables<https://dev.mysql.com/doc/refman/8.0/en/server-status-variables.html#statvar_Open_tables>. Because this feature can keep fixed memory usage, but its performance would be dropped.

Reviewer: 良さそうに思いました！

 <!--ステータス変数の Open_tables <https://dev.mysql.com/doc/refman/8.0/en/server-status-variables.html#statvar_Open_tables> の値とメモリー使用量を見ながら table_open_cache の値を調整することで、メモリー使用量とパフォーマンスのバランスを調整する必要があります。-->

There is how to check `Open_tables` as follows
SHOW GLOBAL STATUS LIKE 'Open_tables';
-- +---------------+-------+
-- | Variable_name | Value |
-- +---------------+-------+
-- | Open_tables   | 643   |
-- +---------------+-------+

> Normally, all of tables in use are cached by setting `table_open_cache` larger than `Open Tables`. However, if there is limited memory, memory usage would be decreased by releasing tables with low usage from the cache.

Author: ```suggestion
Normally, all of tables in use are cached by setting `table_open_cache` larger than `Open Tables`. However, if there is limited memory, memory usage would be decreased by releasing memory for tables with low usage dropping from the cache.
```

説明を伺うとこの表現かなという文章に直しています

In order to releasing tables with low usage from the cache, `table_open_cache` should be set smaller than `Open Tables`. 

> Setting little smaller `Open Tables` more  tables dropping from cache. 

Author: ```suggestion
The little smaller `Open Tables` setting , the more tables dropping from the cache.
```

> As cache is released, memory usage decrease, yet performance become worse.

Author: ```suggestion
As cache is released, memory usage decrease, yet performance become worse due to reopen tables.
```

<!--通常は Open_tables より table_open_cache を大きくして常に使っているすべてテーブルをキャッシュします。 しかし、メモリーに余裕がない環境では一部の使用頻度の低いテーブルをキャッシュから落として解放することでメモリー使用量を削減します。

>  使用頻度の低いテーブルをキャッシュから落とすには Open_tables より table_open_cache を小さくします。 少し小さくすると少しだけキャッシュから落ちます。その分メモリー使用量は減りますが、テーブルの開き直しが必要になるためパフォーマンスは悪化します。 table_open_cache を小さくするほどその度合いは大きくなるので、小さくするのはメモリー使用量が許容範囲内に収まる程度までにしておきます。

Reviewer: 「table_open_cache を小さくするほどその度合いは大きくなるので、小さくするのはメモリー使用量が許容範囲内に収まる程度までにしておきます。」

の部分も翻訳して頂けますか！？

>  使用頻度の低いテーブルをキャッシュから落とすには Open_tables より table_open_cache を小さくします。 少し小さくすると少しだけキャッシュから落ちます。その分メモリー使用量は減りますが、テーブルの開き直しが必要になるためパフォーマンスは悪化します。 table_open_cache を小さくするほどその度合いは大きくなるので、小さくするのはメモリー使用量が許容範囲内に収まる程度までにしておきます。

Author: ```suggestion
The smaller `table_open_cache`, the more necessary to reopen tables. `table_open_cache` should  set as much value as only to reaching memory tolerance.

<!-- table_open_cache を小さくするほどその度合いは大きくなるので、小さくするのはメモリー使用量が許容範囲内に収まる程度までにしておきます。-->
```

ここの日本語も少し英語にしずらかったところでした・・・

その度合いが大きくなる　→パフォーマンス悪化の度合いの話ですか？　それともメモリ使用量の減少の話なのかがちょっと分かりづらかったです。
（メモリー使用量が減るのに、メモリー使用量が許容範囲内に収まるというのがちょっとわからなかったんですが、つまり、テーブル開きなおしが必要になるのに必要なメモリー使用量という理解であっていますか？そのつもりの文章になっているので違ったらご指摘ください・・・）

Reviewer: パフォーマンスの悪化具合の話でした。

メモリー使用量の許容範囲とは、元々この機能はMySQL(Mroonga)がメモリーを使いすぎる場合の対策として使うのですが、ユーザーが思うMySQL(Mroonga)が使って良いメモリー使用量の最大値のことになります。
`table_open_cache` の値を小さくすればするほどMySQL(Mroonga)が使っているメモリー量は減らせるのですが、メモリー使用量が減るのに応じてパフォーマンスが悪くなるので、いい感じに `table_open_cache` を調整して欲しいということになります。

この部分をもう少し丁寧に書くのなら以下のような感じなります。

使用頻度の低いテーブルをキャッシュから落とすには `Open_tables` よりも `table_open_cache` を小さくします。 
`Open_tables` よりも小さくした分だけキャッシュから落ちます。
キャッシュから落ちた分MySQL（Mroonga）のメモリー使用量は減りますが、キャッシュから落ちたテーブルの開き直しが（将来的に）必要になるためパフォーマンスは悪化します。
`table_open_cache` を小さくする程パフォーマンスの悪化の度合いが大きくなるので、 `table_open_cache` を小さくするのは、
MySQL（Mroonga）のメモリー使用量が許容範囲内に収まる程度までにしておきます。

Author: ありがとうございます

メモリーの使用量の許容範囲は tolerance　で大丈夫ですか？

あと、「程度まで」と「しておきます」の日本語から結構緩いのかなと思ったんですけど、
程度の問題の確認なのですが、これは「should」すべき！なのか、recommend 「おすすめ」でいうとどのくらいなんですかね・・・？

いちおう直したのをcommitするので確認してみてください

Reviewer: > メモリーの使用量の許容範囲は tolerance　で大丈夫ですか？

正直技術英語でこういう時に良い表現があるのか分からないのですが、大丈夫そうな感じがしましたので、一旦これで適用しようと思います。

Mroongaは高速全文検索を提供するもので、パフォーマンスの劣化は可能な限り回避したいので、shouldのニュアンスだと思います！

Author: Torelanceは耐久的な許容なのでちょっと違う気がしてっはいて・・・　https://ejje.weblio.jp/content/tolerance

快適に使える範囲のメモリー使用量という意味であれば、"suitable memory usage capacity"　みたいな気もしています

For actual use, we need to adjust balance between memory usage and perfomance with value of `table_open_cache` while checking memory usage and value of static variables Open Tables<https://dev.mysql.com/doc/refman/8.0/en/server-status-variables.html#statvar_Open_tables>. Because this feature can keep fixed memory usage, but its performance would be dropped.

 <!--ステータス変数の Open_tables <https://dev.mysql.com/doc/refman/8.0/en/server-status-variables.html#statvar_Open_tables> の値とメモリー使用量を見ながら table_open_cache の値を調整することで、メモリー使用量とパフォーマンスのバランスを調整する必要があります。-->

There is how to check `Open_tables` as follows
SHOW GLOBAL STATUS LIKE 'Open_tables';
-- +---------------+-------+
-- | Variable_name | Value |
-- +---------------+-------+
-- | Open_tables   | 643   |
-- +---------------+-------+
Normally, all of tables in use are cached by setting `table_open_cache` larger than `Open Tables`. However, if there is limited memory, memory usage would be decreased by releasing memory for tables with low usage dropping from the cache.
In order to releasing tables with low usage from the cache, `table_open_cache` should be set smaller than `Open Tables`. 

> The little smaller `Open Tables` setting , the more tables dropping from the cache.

Reviewer: ここのlittle smallerのlittleのニュアンスはここではいらない気がするので、適用時に削除しておきます。

Author: 了解です！
